File: distros/J/JavonetPerlSdk/lib/Javonet/Binaries/Ruby/Linux/X64/core/handler/set_instance_field_handler.rb
require_relative 'abstract_command_handler'

class SetInstanceFieldHandler < AbstractCommandHandler
  def initialize
    @required_parameters_count = 3
  end

  def process(command)
    return set_instance_field(command)
  end

  def set_instance_field(command)
    begin
      if command.payload.length < @required_parameters_count
        raise ArgumentError.new "Set Instance Field parameters mismatch"
      end
  
      merged_value = '@' + command.payload[1]
      begin
        command.payload[0].instance_variable_set(merged_value, command.payload[2])
      rescue NameError
        fields = command.payload[0].instance_variables
        message = "Field #{command.payload[1]} not found in class #{command.payload[0].class.name}. Available fields:\n"
        fields.each { |field_iter| message += "#{field_iter} \n" }
        raise Exception, message
      end
    rescue Exception => e
      return e
    end
  end
end

File: distros/J/JavonetPerlSdk/lib/Javonet/Binaries/Nodejs/core/handler/SetInstanceFieldHandler.js
const AbstractHandler = require("./AbstractHandler");

class SetInstanceFieldHandler extends AbstractHandler {
  requiredParametersCount = 3;

  constructor() {
    super();
  }

  process(command) {
    try {
      if (command.payload.length < this.requiredParametersCount) {
        throw new Error("Set Instance Field parameters mismatch");
      }
      const { payload } = command;
      let instance = payload[0];
      let field = payload[1];
      let value = payload[2];
      if (typeof instance[field] === "undefined") {
        let fields = Object.keys(instance);
        let message = `Field ${field} not found in object. Available fields:\n`;
        fields.forEach((fieldIter) => {
          message += `${fieldIter}\n`;
        });
        throw new Error(message);
      }
      instance[field] = value;
      return 0;
    } catch (error) {
      throw this.process_stack_trace(error, this.constructor.name);
    }
  }
}

module.exports = new SetInstanceFieldHandler();
